fix(common): date is not correctly formatted when year is between 0 and 99

abarghoud opened the pull request:
…nd 99

use setFullYear method when parsing date to avoid javascript date factory behaviour, which converts year between 0 and 99 to 1900 - 1999

Fixes #40377

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

abarghoud committed:
fix(common): date is not correctly formatted when year is between 0 and 99

use setFullYear method when parsing date to avoid javascript date factory behaviour

Fixes #40377

petebacondarwin commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -363,14 +363,22 @@ const JANUARY = 0;
 const THURSDAY = 4;
 function getFirstThursdayOfYear(year: number) {
   const firstDayOfYear = (new Date(year, JANUARY, 1)).getDay();
```
You will need to change this too I believe, since this is used to compute the first Thursday.
In which case, perhaps write a helper function `createDate()` that hides the need to call `setFullYear()`?

abarghoud replied:
Yes ! you are right.

abarghoud replied:
Actually @petebacondarwin after thinking about the helper function I think it will a bit complicated to write this function, because if I do so, it should react, accept and have the same logic as the Date constructor, what do you think ?

Regarding the need to do the same for `firstDayOfYear` it's not really needed as we create this date object only to get the day, so the year here is not a problem.

petebacondarwin replied:
The result of `getDay()` changes depending upon the current year:

```
> x = new Date(1, 1, 1)
1901-02-01T00:00:00.000Z
> x.getDay()
5
> y = new Date(1, 1, 1)
1901-02-01T00:00:00.000Z
> y.setFullYear(1)
-62132918325000
> y.getDay()
4
```

petebacondarwin replied:
Regarding the helper, I was only thinking of an internal function, avoiding repeating the code that sets the full year:

```ts
function createDate(year: number, month: number, date: number): Date {
  const date = new Date(year, month, number);
  date.setFullYear(year);
  return date;
}
```

abarghoud replied:
Oh, ok I see what you mean

abarghoud replied:
@petebacondarwin done

petebacondarwin commented:
Also note that the commit body lines must not exceed 120 chars: https://app.circleci.com/pipelines/github/angular/angular/27109/workflows/1258d805-3b40-4e69-abc6-947b78ac87cb/jobs/897541

abarghoud commented:
@petebacondarwin I will fix the tests, actually I have a small problem with test execution and clang-format locally because I use a macbook with an M1 chip, so puppeteer and clang-format cannot be installed unfortunately,.

petebacondarwin commented:
Don't worry about running the tests in a browser. The node only tests should be enough for you locally.

abarghoud commented:
> Don't worry about running the tests in a browser. The node only tests should be enough for you locally.

@petebacondarwin, would you mind giving me the command to run only node tests, I can't find it here https://github.com/angular/angular/blob/master/docs/DEVELOPER.md#running-tests-locally

petebacondarwin commented:
You can see the various tests that are run in the CircleCI logs. The one that is relevant to you here is `//packages/common/test:test`. You can run that by:

```
yarn bazel test //packages/common/test
```

(the `:test` is not needed because if not provided Bazel just uses the last segment of the path)

petebacondarwin commented on packages/common/test/i18n/format_date_spec.ts:
```diff
@@ -378,5 +378,11 @@ describe('Format date', () => {
+
+    // https://github.com/angular/angular/issues/40377
+    it('should format date with year between 0 and 99 correctly', () => {
+      expect(formatDate('0099-01-11T12:00:00.014-0500', defaultFormat, ɵDEFAULT_LOCALE_ID))
```
Does this test actually fail without the fix?
https://stackblitz.com/edit/angular-ivy-qhzmxz?devtoolsheight=33&file=src/app/app.component.ts

petebacondarwin replied:
I think you actually don't want to pass in an ISO formatted date, since that will actually get parsed correctly!!!  Instead this should be:

```suggestion
      expect(formatDate('0099-01-11', defaultFormat, ɵDEFAULT_LOCALE_ID))
```

See https://stackblitz.com/edit/angular-ivy-mxrnbk

abarghoud replied:
Actually it fails when you use `"YYYY"` as format, I will change `defaultFormat` to `"YYYY"`

abarghoud replied:
https://stackblitz.com/edit/angular-ivy-lbeum3?devtoolsheight=33&file=src/app/app.component.ts

petebacondarwin commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -101,6 +101,18 @@ export function formatDate(
+ */
+function createDate(year: number, month: number, date: number): Date {
+  const fullYearDate = new Date(year, month, date);
+  if (year >= 0 && year <= 99) {
```
Is there any downside in always setting the full year?

The performance cost of writing it unnecessarily is probably about the same as the cost of testing the year twice.

I would imagine it is not necessary, and that removing this check would keep the size increase lower.

abarghoud replied:
Actually when not having this check, this test fails https://github.com/angular/angular/pull/40448/files#diff-104f5ff9c7d7b28cabe175a2c5dbc49e3399212e863d78c5f3e4d79e5983a1d8R374

```
it('should return correct ISO 8601 week-numbering year for dates close to year end/beginning',
       () => {
         expect(formatDate('2013-12-27', 'YYYY', 'en')).toEqual('2013');
         expect(formatDate('2013-12-29', 'YYYY', 'en')).toEqual('2014');
         expect(formatDate('2010-01-02', 'YYYY', 'en')).toEqual('2009');
         expect(formatDate('2010-01-04', 'YYYY', 'en')).toEqual('2010');
       });
```

petebacondarwin replied:
Hmmm, that is a bit worrying. Can you check that this test passes for dates in the [0...99] range?

abarghoud replied:
will do

abarghoud replied:
I tested it and actually it fails when date is 01-02 in which case it should be a year before https://app.circleci.com/pipelines/github/angular/angular/27129/workflows/6792f442-04b5-45b9-87c2-df9e03513aeb/jobs/897893/parallel-runs/0/steps/0-106

petebacondarwin replied:
So those cases should be included in our tests; and we should make the code work for them.
I wonder what the problem is with calling `setFullYear()` in these cases?

abarghoud replied:
Yes I agree, I think that setFullYear modifies the year calculated in weekNumberingYearGetter method, which has been added in https://github.com/angular/angular/issues/38739

abarghoud replied:
I identified the problem i will be working on a fix

petebacondarwin replied:
👍

abarghoud replied:
Resolved

AndrewKushnir commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -101,6 +101,18 @@ export function formatDate(
+function createDate(year: number, month: number, date: number): Date {
+  const newDate = new Date(year, month, date);
+  if (year >= 0 && year <= 99) {
+    newDate.setFullYear(year - (year - (newDate.getFullYear() - 1900)));
```
Wouldn't it be just this (year - year = 0)?
```suggestion
    newDate.setFullYear(newDate.getFullYear() - 1900);
```

abarghoud replied:
yes it was stupid 😅, done

AndrewKushnir commented on packages/common/test/i18n/format_date_spec.ts:
```diff
@@ -377,6 +377,13 @@ describe('Format date', () => {
+
+    // https://github.com/angular/angular/issues/40377
+    it('should format date with year between 0 and 99 correctly', () => {
+      expect(formatDate('0099-01-11', 'YYYY', ɵDEFAULT_LOCALE_ID)).toEqual('0099');
```
I'd propose to add a couple more checks to test borderline conditions and how we transition in/out the [0, 99] range, i.e. test with the following years:

- 0100
- 0099
- 0098
- 0001
- 0000

abarghoud replied:
done

petebacondarwin commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -101,6 +101,18 @@ export function formatDate(
+ * https://github.com/angular/angular/issues/40377
+ */
+function createDate(year: number, month: number, date: number): Date {
+  const newDate = new Date(year, month, date);
+  if (year >= 0 && year <= 99) {
+    newDate.setFullYear(newDate.getFullYear() - 1900);
+  }
+  return newDate;
```
This can actually be simplified:

```suggestion
  const newDate = new Date(0);
  newDate.setFullYear(year, month, date);
  return newDate;
```

The reason that it didn't work before is that if you do not specify month and date in `setFullYear()` it retrieves the values from the current date via `getMonth()` and `getDate()`, which get messed up with locale timezone offsets etc.

abarghoud replied:
Yes exactly, done

petebacondarwin reviewed (approved):
Thanks for iterating on this @abarghoud. LGTM now.

AndrewKushnir commented:
[Presubmit](http://test/OCL:352212686:BASE:352212684:1610842095943:6be6cf1c).

AndrewKushnir commented:
@abarghoud, @petebacondarwin presubmit indicated that there are some failing targets and I've extracted the repro, [you can find it here](https://stackblitz.com/edit/angular-ivy-repro-40448?file=src/app/app.component.html). The error message in an internal failing test says:
```
Expected 'Sep 20, 2019, 4:00:00 PM UTC-07:00' to be 'Sep 20, 2019, 12:00:00 AM UTC-07:00'.
```
Could you please try to repro that as a test case using the mentioned information? Please let me know if more investigation/information is needed, I can try to work on exact test case.

I'm also marking this PR as a "breaking change" and added a "blocked" label, we'll update the labels (and possibly remove both of them) as needed once there is more info after additional investigation.

Thank you.

abarghoud commented:
> @abarghoud, @petebacondarwin presubmit indicated that there are some failing targets and I've extracted the repro, [you can find it here](https://stackblitz.com/edit/angular-ivy-repro-40448?file=src/app/app.component.html). The error message in an internal failing test says:
> ```
> Expected 'Sep 20, 2019, 4:00:00 PM UTC-07:00' to be 'Sep 20, 2019, 12:00:00 AM UTC-07:00'.
> ```
> Could you please try to repro that as a test case using the mentioned information? Please let me know if more investigation/information is needed, I can try to work on exact test case.
> 
> I'm also marking this PR as a "breaking change" and added a "blocked" label, we'll update the labels (and possibly remove both of them) as needed once there is more info after additional investigation.
> 
> Thank you.

@AndrewKushnir Thank you for the information, i will look into it.

I am a bit curious to know more about presubmit, Are there different tests executed than the ones on the CI ?

petebacondarwin commented:
@abarghoud - the projects that use Angular internally all use the same version: the current master branch! In order to ensure that our commits do not break these 1000s of applications we run the tests of a sample of projects before we merge into master. This is called "G3 presubmit". It is a very good test of breaking changes in our code!!

abarghoud commented:
@AndrewKushnir @petebacondarwin, Actually I wrote a test that reproduces the failing test but I got a different failing result than what you posted, the result I got is: 
` Expected 'Sep 19, 2019, 5:00:00 PM UTC-07:00' to equal 'Sep 20, 2019, 12:00:00 AM UTC-07:00'.`
and I think that it's the correct behaviour as we set the timezone to `-7` so `Sep 20, 2019, 12:00:00 AM UTC` `-7` equals `Sep 19, 2019, 5:00:00 PM UTC-07:00`

I also checked the stackblitz that you sent, it gives `Date: Sep 20, 2019, 12:00:00 AM UTC+02:00` but when defining timezone to `-7` it become `Date: Sep 19, 2019, 3:00:00 PM UTC-07:00` which is correct too giving browser timezone is `UTC+02:00`

petebacondarwin commented:
and of course, I get a different result again, since I am in a different timezone to either of you:

```
Date: Sep 19, 2019, 4:00:00 PM UTC-07:00
```

So I think the point is that, previously, our date-time parsing was locale specific, whereas this PR makes the parsing locale agnostic. From looking at our docs, we claim that passing a string to `DatePipe`/`formatDate()` is evaluated as an [ISO Date-time string](https://angular.io/api/common/formatDate), which is defined in the [ECMAScript specification](https://tc39.es/ecma262/#sec-date-time-string-format). But the spec doesn't say what the default times should be for date only strings.

Looking into this further I found the [MDN doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/parse#date_time_string_format) which states:

>  When the time zone offset is absent, date-only forms are interpreted as a UTC time and date-time forms are interpreted as local time.

So if we were to follow that, then `2019-09-20` should be interpreted as UTC (as in this PR) but if any time is included - e.g. `2019-09-20T00:00:00` then it should default to the local time (assuming no timezone value was provided).

---

I think this means that this PR is actually a breaking change - but a valid one, that we should look to roll out for v12. I don't know how much impact this would have in G3...

AndrewKushnir commented:
> I don't know how much impact this would have in G3...

@petebacondarwin I've started a [global presubmit](https://test.corp.google.com/OCL:352212686:BASE:352495253:1611036411902:9e30499b) (to test a wider set of internal application) to see the impact. Will keep this thread updated.

abarghoud commented:
@petebacondarwin , so in this case should I fix the test that I wrote to reproduce the failing test from presubmit ?

petebacondarwin commented:
Before you work on this further, let's wait for the results from the global presubmit. It might be that despite this being the "correct" solution, we can't land it because it would be too breaking. Also keen to hear from both you (@abarghoud) and @AndrewKushnir on whether you agree that this is indeed "correct".

abarghoud commented:
> Before you work on this further, let's wait for the results from the global presubmit. It might be that despite this being the "correct" solution, we can't land it because it would be too breaking. Also keen to hear from both you (@abarghoud) and @AndrewKushnir on whether you agree that this is indeed "correct".

Yes I totally agree that this is the correct behaviour as I said here https://github.com/angular/angular/pull/40448#issuecomment-762441339 and even in the code it's obvious when we call `convertTimezoneToLocal` https://github.com/angular/angular/blob/master/packages/common/src/i18n/format_date.ts#L90 which adds the offset of timezone to the date.

AndrewKushnir commented:
Thanks for looking into this @abarghoud and @petebacondarwin! I agree that the new behavior looks more inline with specs, so the change is correct. We should make sure we document it extensively so that this information is clear and easily discoverable (we can update API docs as well as include that info into corresponding guide(s)). I'd also propose adding more tests that would specifically exercise the codepath that might produce different outcome based on the date+time vs just date string.

Re: global presubmit: good news, this change affects less than 10 targets, which should be feasible to update/cleanup.

petebacondarwin commented:
Thanks for the update @AndrewKushnir. Are you able to tell if the failures in the presubmits were just tests that had been setup to match the current behaviour, or whether it would actually cause a problem in the running app?

AndrewKushnir commented:
@petebacondarwin, the issues are mostly in tests, but there is a couple failures in app logic, one of them passes params to a function but applies date pipe transformation before doing that, i.e. something like this:
```
{{ someFn(dateField | date: 'yyyy-MM-dd':'+0000') }}
```
Also most of other failures have similar origin: while processing date (without time part), there are expectations that the time would be set as midnight (similar to what I mentioned before). I believe this is something that other people might expect as well...

After looking at the test cases and apps code, I feel like the breaking change might be relatively substantial with arguably less expected new behavior. Would it be possible to keep existing behavior and just fix the original issue (for 0-99 range)?

petebacondarwin commented:
@AndrewKushnir - according to the Angular docs this:

```
{{ someFn(dateField | date: 'yyyy-MM-dd':'+0000') }}
```

is saying that the date should be treated with timezone of `+0`, which is UTC, rather than the local system timezone.

This seems to me exactly what we are doing in this PR. In other words, I don't understand why that particular application would fail in this case since the caller is specifically asking for a UTC date...

Or am I missing something?

petebacondarwin commented:
Ah - I see there is a difference... I downloaded that Stackblitz locally and used the artifacts from this PR. The result (for Mountain View) is:

```
2021-01-23 - 00:00:00 +0000

2021-01-22 - 16:00:00 -0800
```

Note that the date has changed from 22 to 23...

petebacondarwin commented:
Oh! I just realised that our own CI tests are failing! So it is not that surprising that global presubmit is too!

abarghoud commented:
> Oh! I just realised that our own CI tests are failing! So it is not that surprising that global presubmit is too!

@petebacondarwin Actually it's failing because I've added a test for this case https://github.com/angular/angular/pull/40448#issuecomment-761880672 to make sure to have that case covered in Angular tests, I did not fix it because as you told me here https://github.com/angular/angular/pull/40448#issuecomment-762723263 we should wait to have some more feedback from the global presubmit

petebacondarwin commented:
Doh! Sorry @abarghoud - yes I remember now. Looking at that test, though, I think it will always fail outside of a specific region if when we parse `2019-09-20` is taken as a local date time.

petebacondarwin commented on packages/common/test/i18n/format_date_spec.ts:
```diff
@@ -377,6 +379,17 @@ describe('Format date', () => {
          expect(formatDate('2013-12-29', 'YYYY', 'en')).toEqual('2014');
          expect(formatDate('2010-01-02', 'YYYY', 'en')).toEqual('2009');
          expect(formatDate('2010-01-04', 'YYYY', 'en')).toEqual('2010');
+         expect(formatDate('0049-01-01', 'YYYY', 'en')).toEqual('0048');
```
By the way, according to https://www.timeanddate.com/date/weeknumber.html the week number year in this case should be 49.

<img width="353" alt="Screenshot 2021-01-22 at 10 04 36" src="https://user-images.githubusercontent.com/15655/105477085-52923f80-5c99-11eb-87f6-1d3f7169f02e.png">

If we choose 0046 then we get a week that starts the previous year:

<img width="378" alt="Screenshot 2021-01-22 at 10 06 01" src="https://user-images.githubusercontent.com/15655/105477219-7ce3fd00-5c99-11eb-9aa2-fd9e8404aa19.png">

But I might well be missing something (again!).

abarghoud replied:
I think it's again a location issue, as when I do the same test I am getting `30 Dec 0045` as Start Date:
![image](https://user-images.githubusercontent.com/13522017/105478357-5fb42c00-5ca3-11eb-86f1-18d300887482.png)

petebacondarwin replied:
Urgh! What a mess this is.

abarghoud replied:
Yes javascript date is a mess 😅

abarghoud commented:
> Ah - I see there is a difference... I downloaded that Stackblitz locally and used the artifacts from this PR. The result (for Mountain View) is:
> 
> ```
> 2021-01-23 - 00:00:00 +0000
> 
> 2021-01-22 - 16:00:00 -0800
> ```
> 
> Note that the date has changed from 22 to 23...

@petebacondarwin So I tried to reproduce what you described above in a unit test 
https://github.com/abarghoud/angular/blob/correctly-format-dates-between-0-99/packages/common/test/i18n/format_date_spec.ts#L338-L341

And as you can see in the CI https://app.circleci.com/pipelines/github/angular/angular/27381/workflows/99788338-cc2e-4212-8272-f0801988ecc6/jobs/902130, the results are correct:

```
expect(formatDate('2021-01-22', 'yyyy-MM-dd - HH:mm:ss Z', ɵDEFAULT_LOCALE_ID))
          .toEqual('2021-01-22 - 00:00:00 +0100');
// execution result: Expected '2021-01-22 - 00:00:00 +0000' to equal '2021-01-22 - 00:00:00 +0100'

expect(formatDate('2021-01-22', 'yyyy-MM-dd - HH:mm:ss Z', ɵDEFAULT_LOCALE_ID, '+0000'))
          .toEqual('2021-01-21 - 23:00:00 +0000');
// execution result: Expected '2021-01-22 - 00:00:00 +0000' to equal '2021-01-21 - 23:00:00 +0000'.
```

abarghoud commented:
> @petebacondarwin, the issues are mostly in tests, but there is a couple failures in app logic, one of them passes params to a function but applies date pipe transformation before doing that, i.e. something like this:
> 
> ```
> {{ someFn(dateField | date: 'yyyy-MM-dd':'+0000') }}
> ```
> 
> Also most of other failures have similar origin: while processing date (without time part), there are expectations that the time would be set as midnight (similar to what I mentioned before). I believe this is something that other people might expect as well...
> 
> After looking at the test cases and apps code, I feel like the breaking change might be relatively substantial with arguably less expected new behavior. Would it be possible to keep existing behavior and just fix the original issue (for 0-99 range)?

Hello @AndrewKushnir, In order to make this fix and keep existing behavior, I would like you to confirm to me the current behaviors that we've identified until this moment, so I can write some unit tests that covers them in the code:

1. There is one here https://github.com/angular/angular/pull/40448#issuecomment-761880672 which we've considered as a valid breaking change, would it still be considered as a valid behavior ?
2. The other one is https://github.com/angular/angular/pull/40448#issuecomment-765076140 which I cannot reproduce, I've added some unit tests to test that giving a date without time part the time should be set to midnight, and these tests are passing.

@petebacondarwin @AndrewKushnir what do you think ?

petebacondarwin commented:
I guess we need to decide whether a date-only string should be parsed as UTC-0 or parsed as though it is midnight in the local timezone.

The `Date()` constructor, which I believe works the same way as `Date.parse()` as string that contains no time parts should be treated as UTC.See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date#Timestamp_string

> Support for ISO 8601 formats differs in that date-only strings (e.g. "1970-01-01") are treated as UTC, not local.

So I think that this breaking change really is what we want. I think the workaround for apps that expect it to be local is to add a time to the string. For example:

```ts
formatDate('2019-09-20T00:00:00')
```

petebacondarwin committed:
fixup! fix(common): date is not correctly formatted when year is between 0 and 99

petebacondarwin committed:
fixup! fix(common): date is not correctly formatted when year is between 0 and 99

petebacondarwin committed:
fixup! fix(common): date is not correctly formatted when year is between 0 and 99

petebacondarwin commented:
OK, so I spent some more time on this today.

I think we can indeed avoid a breaking change (even though I believe making one would be more inline with how JS parsing of ISO date strings works).

I have added a fixup commit, which I believe allows this fix to work without breaking local time zone requirements for backward compatibility. I have tested this locally on Chrome by setting the timezone to a variety of locations. The way the tests are setup ensures that they pass whichever time zone the tests are being run from.

@abarghoud and @AndrewKushnir can you take a look and let me know what you think?

petebacondarwin commented:
@googlebot I consent.

abarghoud commented on packages/common/test/i18n/format_date_spec.ts:
```diff
@@ -347,6 +347,32 @@ describe('Format date', () => {
+    //    });
+
+    // This test is to ensure backward compatibility for parsing date-only ISO strings.
+    it('should create local timeezone date objects when an ISO string is passed with no time components',
```
```suggestion
    it('should create local timezone date objects when an ISO string is passed with no time components',
```

abarghoud reviewed (commented):
@petebacondarwin A small typo, otherwise LGTM

abarghoud commented:
> OK, so I spent some more time on this today.
> 
> I think we can indeed avoid a breaking change (even though I believe making one would be more inline with how JS parsing of ISO date strings works).
> 
> I have added a fixup commit, which I believe allows this fix to work without breaking local time zone requirements for backward compatibility. I have tested this locally on Chrome by setting the timezone to a variety of locations. The way the tests are setup ensures that they pass whichever time zone the tests are being run from.
> 
> @abarghoud and @AndrewKushnir can you take a look and let me know what you think?
> 
> (Oh and I rebased and force-pushed the branch @abarghoud - hope that doesn't mess up any local work you had on this branch.)

I think that we should modify the docs here https://angular.io/api/common/DatePipe#custom-format-options to indicate that time will be set to midnight, because currently it only says 

> Note that timezone correction is not applied to an ISO string that has no time component, such as "2016-09-19"
@petebacondarwin what do you think ?

petebacondarwin commented:
> I think that we should modify the docs here https://angular.io/api/common/DatePipe#custom-format-options to indicate that time will be set to midnight, because currently it only says
> 
> > Note that timezone correction is not applied to an ISO string that has no time component, such as "2016-09-19"

Well that is an interesting statement in the docs :-) Basically that is saying that parsing date-only strings should NOT use local date. Clearly that appears not to be the case. So indeed we should remove that statement.

This text was basically added (but then later modified) in https://github.com/angular/angular/commit/2aba8b0ff2c5d6979802b96842aceccad3d5716d

So I think we should consider a breaking change to re-fix https://github.com/angular/angular/issues/12373 in the future...

abarghoud commented:
> @abarghoud - do you want to update this PR with an additional documentation fix?

@petebacondarwin Yes I will, so if I understand correctly we should only remove this statement without adding another one to explain the behaviour ?

petebacondarwin commented:
At the least remove it. But probably some explanation of the current behaviour would be better

AndrewKushnir commented:
[Presubmit](http://test/OCL:356283063:BASE:356283081:1612805362403:f5908a0d).

AndrewKushnir commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -101,6 +101,17 @@ export function formatDate(
+function createDate(year: number, month: number, date: number): Date {
+  const newDate = new Date(0);
+  newDate.setFullYear(year, month, date);
+  newDate.setHours(0, 0, 0);
```
I'd propose to add a quick comment on why we've added this line. Function description provides very good context and we can also simplify further work on this (once we decide to do a breaking change) by making it more specific that this is the part of code we should re-consider in the future.

petebacondarwin replied:
That's fair enough. I'll add another fixup commit.

petebacondarwin replied:
Done

AndrewKushnir replied:
Awesome comments, thanks Pete! 👍

abarghoud commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -101,6 +101,37 @@ export function formatDate(
+  // (Note that they even have different years, dates and months.)
+  const newDate = new Date(0);
+
+  // `setFullYear()` allows years like 1901 to be set correctly. This function does not
```
@petebacondarwin Because 1901 can be set correctly using `new Date(0)`, can't be ?
```suggestion
  // `setFullYear()` allows years like 0001 to be set correctly. This function does not
```

petebacondarwin replied:
Doh!! Good spot @abarghoud  :-)

AndrewKushnir commented:
FYI, presubmits (including global one) are successful for the changes in this PR.

It looks like this change is not breaking anymore, we should consider changing the target to "patch".

Thank you.

petebacondarwin commented:
Yes! Let's go with patch. Thanks @AndrewKushnir
